deepCopy also copies type information of lists

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23271 deepCopy also copies type information of lists**

Differential Revision: [D16449220](https://our.internmc.facebook.com/intern/diff/D16449220/)

smessmer committed:
deepCopy also copies type information of lists



Differential Revision: [D16449220](https://our.internmc.facebook.com/intern/diff/D16449220/)

dzhulgakov commented on torch/csrc/jit/passes/utils/check_alias_annotation.cpp:
```diff
@@ -25,12 +25,22 @@ IValue deepCopy(const IValue& self) {
+      }
+    };
+    if (source._elementType().has_value()) {
+      auto newList = c10::impl::GenericList(*source._elementType());
```
can it be done without lambda?

```
auto newList = source._elementType().has_value() ? c10::impl::GenericList(*source._elementType()) : c10::impl::GenericList(c10::impl::deprecatedUntypedList());
...
```

smessmer replied:
I hadn't thought of the conditional expression. Thanks, will use that instead.

smessmer committed:
Update on "deepCopy also copies type information of lists"


Differential Revision: [D16449220](https://our.internmc.facebook.com/intern/diff/D16449220/)
